### PR TITLE
Implement tag-encoding to avoid npm errors

### PR DESCRIPTION
**Pull request by minkimcello**
## Motivation
Part of our publishing process involves publishing pull request packages to NPM using branch names as tags. This allows us to install and test new changes of a pull request without having to wait for approval & merge. However, installing packages with tags that have slashes in them confuses the NPM CLI so this PR is to introduce a workaround.

### npm-publish-branch-preview
The `thefrontside/actions/npm-publish-branch-preview` action will publish to NPM with the head branch name as its tag. The new proposed changes will encode the branch name so that all instances of `/` are replaced with `_` as well as adding an additional `_` after each group of `_` if branch already has those characters in its name (i.e. one underscore would become two; four underscores would become five).

Examples:
- `mk/branch_name` will be encoded as `mk_branch__name`
- `mk_branch__name` will become `mk__branch___name`

### synchronize-npm-tags
This action has also been updated to adapt to the tag-encoding. This action, which runs on our `on-delete` workflow, will retrieve all the tags from NPM (which are already encoded from `npm-publish-branch-preview`) to cross check them with branch names of the Github repository; it's so that we can remove unnecessary NPM tags once a branch is deleted.

With the new update, this action will encode branch names and user-specified argument (to preserve tags that do not have corresponding branches) in the same manner as how it was encoded in `npm-publish-branch-preview`.

## Conclusion
In summary, this encoding only happens before publishing tags *to* NPM and as it makes checks to see which ones to remove *from* NPM so it should not cause any issues elsewhere.

### CODE REVIEW

**Comment by minkimcello** [comment]
@cowboyd Just occurred to me that because we're adding `_` to each groups of `_`, we need to decide if the argument (to preserve tags) should be in the format of branch names or npm tag labels.

File: npm-publish-branch-preview/entrypoint.sh [hunk 1]
```diff
@@ -23,6 +23,6 @@ else
   echo "//registry.npmjs.org/:_authToken=$NPM_AUTH_TOKEN" > ~/.npmrc
   npm config set unsafe-perm true
   npm install
-  tag="$(echo $GITHUB_HEAD_REF | sed -E 's/\_+/&\_/g;s/\//\_/g')"
+  tag="$(echo $GITHUB_HEAD_REF | sed -E 's/\_/\_\_/g;s/\//\_/g')"
```
Comment (cowboyd) [inline npm-publish-branch-preview/entrypoint.sh]: I don't think that `_` is a special character in `sed` so shouldn't need to be escaped. The only thing that needs to be escaped is the `/` because it's used as the delimiter so 

`s/_/__/g;s/\//_/g`

ought to work. But since we're explicitly matching on `/` it might be better to use a different delimiter than the conventional `/` like `:`

`s:_:__:g;s:/:_:g;`

Comment (minkimcello) [reply npm-publish-branch-preview/entrypoint.sh]: Are the different delimiters just to make everything more legible or do they have different functionality?

Comment (cowboyd) [reply npm-publish-branch-preview/entrypoint.sh]: Nope, same functionality. Generally you use different delimiters when your matching directly involves the conventional delimiter `/`
